to-fix backend authentication

Rub21 opened the pull request:
cc. @rodowi @ajithranka @batpad @aaronlidman @geohacker

Commit:
to-fix auth

Commit:
fix testing in auth

Commit:
add conf and routes

Commit:
authentication and roles

defvol commented:
> cc. @rodowi @ajithranka @batpad @aaronlidman @geohacker

@Rub21 could you please give details about the feedback you need?

ajithranka commented on index.js:
```diff
@@ -37,11 +41,30 @@ var good = {
+      expiresIn: 24 * 60 * 60 * 1000
+    },
+    cookieOptions: {
+      password: process.env.Password || 'abcdefghigklmnopqrsdasdasdadadadsdtuvwxyz123456',
```
@Rub21 @geohacker Not sure about the context here but is it okay to have a hard-coded password?

Commit:
add index in congif and autentication to update a role

Commit:
autentication to delete a task

Commit:
cleanup

Commit:
update auth for dev, staging and prod

Rub21 commented:
@rodowi @ajithranka @batpad @aaronlidman @geohacker @planemad @maning 
I am working  in the backend authentication:  there already has some progress here:

How is it working?

- it is using grant use OSM third authentication.
- When a user is authenticated in to-fix. to-fix-backend  save the information of the user in ES:

```
{
  id: "1781294",
  user: "Rub21_nycbuildings",
  img: "http://api.openstreetmap.org/attachments/users/images/000/510/836/original/a627791e1bd2ff94ec5b7c40860f57b3.jpg",
  role: "editor"
}
```

you can start the authentication in :https://build-to-fix-staging.tilestream.net/connect/openstreetmap it will redirect to osm login page and grant access, and then it will redirect  to https://build-to-fix-staging.tilestream.net/handle_openstreetmap_callback with  user information : 

```
{
id: "1781294",
user: "Rub21_nycbuildings",
img: "http://api.openstreetmap.org/attachments/users/images/000/510/836/original/a627791e1bd2ff94ec5b7c40860f57b3.jpg",
role: "editor"
}
```

when someone is authenticated at the first time in to-fix it will have per default the editor role.

- We are using three roles, `superadmin`, `admin`, `editor`:

 
  *Superadmin*
    - Can change the role of a user  to admin or superadmin
    - Can create any task
   - Can delete any task

  *admin*
    - can create  tasks.
    - can remove his own task

  *editor*
    - can access to to-fix to start fixing the editions, as we are doing currently 

- Superadmin is creating  for the first user who authenticated in an instance(stack), after that this superadmin can change roles for other users

Check all users: https://build-to-fix-staging.tilestream.net/users
All endpoint here: https://build-to-fix-staging.tilestream.net/docs


what is missing?

-Activate the authentication to create, update and delete a task.

- Testing is becoming painful fro the authentication, anybody recommendations? I am trying the using `nock`, but I need to figure out more,  it is not a simple response from OSM API, it has many redirects and it should return a valid token, wich we use to get the user information and register in the database.

Commit:
remove wrong endpoint

Commit:
fix request client

ajithranka commented:
Hi @Rub21! This looks great.

Can we also think of a way to limit tasks to certain users? So when a `admin` or `superadmin` creates a new task, he might assign a list of users who can see and edit the task. This will let us create private tasks that are only visible to a group of users.

I'm not sure how difficult this will be to implement though. After logging in, each call to the API will have to pass the user details. End points for tasks, stats and activities will have to check on the back-end if the user has access to those tasks.

Let me know if you have any thoughts around this. An alternative could be to create a private instance instead for all private tasks.

Commit:
merge from master last stats changes

Commit:
redirect to client side

Commit:
issecure for sesion

Commit:
logout

Commit:
empty

defvol commented:
great work @Rub21! your approach makes sense

A couple of notes:
- I tried to login on staging but got redirected to a localhost url
https://localhost:3000/oauth-complete.html?id=2334596...&role=editor. We might need to allow for custom hosts, so the base url can be loaded dynamically.
- Can we make the /users endpoint private? Not sure if it's cool to show a bunch of users profiles on a public endpoint

defvol commented:
> Testing is becoming painful fro the authentication, anybody recommendations? I am trying the using nock, but I need to figure out more, it is not a simple response from OSM API, it has many redirects and it should return a valid token, wich we use to get the user information and register in the database.

🤔 a couple of thoughts here:

1. we could skip testing the oauth roundtrip and mock a token to test whichever method evaluates the response from the OSM API
2. using nock we can fake each step of the auth handshake, i.e. use nock to return a 302 response to another fake url where nock will return a fake token.

Commit:
start using jwt2 and generate access token

Commit:
remove user and fix bugs

Commit:
remove password to create and update  a task

Commit:
update auth and remove logout

Commit:
fix commnet

Commit:
fix testing with token

Commit:
redirect to host

Commit:
redirect to client and comment stats test

Commit:
fix missing module

Commit:
add user details and auth for /users endpoint

Commit:
increase the exp time for superadmin

Commit:
debugin test

Commit:
auth in tape test

Commit:
comment auth testing

Commit:
merge from master

Commit:
add stagingside

Commit:
add logout

Commit:
uncomment access to /users

ajithranka commented:
Hi @Rub21, this looks 👍! 

@geohacker, @kepta and I spent some time going through the code. We think there's a few things that might need to be checked before we merge.

- The redirect for production [here](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/configs/config.json#L33) should be `https://osmlab.github.io/to-fix/oauth-complete.html`. 😅 
- Is it ok to have these hard-coded passwords [here](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/index.js#L54) and [here](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/index.js#L69).
- [Creating new tasks](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/routes/routesTasks.js#L31), updating or deleting tasks requires `superadmin` or `admin` privileges. Do we need the `machine` role to also be able to handle tasks?
- Should we have a strict [check on uploaded files](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerTasks.js#L93) being `.geojson`? The file could be a `.json` as well. 
- Do we need to catch errors [here](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerTasks.js#L134-L136)?
- When updating a task, `isCompleted` should not be [mandatory](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/routes/routesTasks.js#L64
). Also, it shouldn't be [hard-coded in the front-end](https://github.com/osmlab/to-fix/blob/master/src/components/admin/edit_task.js#L65).
- The block of code starting from [here](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerTasks.js#L223) which reads the uploaded file, validates it and updates the database could be broken into a function. It's repeated in both `createTask` and `updateTask`.
- Same thing with [this block](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerTasks.js#L288).
- Some checks might be useful when parsing the OSM [user XML](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerUsers.js#L75). This might break, for instance, if `result.osm` is empty.
- We probably need to [remove this](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerUsers.js#L109-L112). Sets the user role to `superadmin` if the response is empty.
- Do we need this [`userExists`](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerUsers.js#L124~) global? Can we limit it to the callback?
- Grammatical change in [this error message](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerUsers.js#L171): `Bad authentication` -- > `Authentication failure`.
- [Get user](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerUsers.js#L185-L205) by id instead of username. Usernames can change and using string match on elastic search can be avoided (it will return many results). Same with [changeRole](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerUsers.js#L228-L268). **We should avoid relying on usernames everywhere. Let's use user ids as far as possible. This will require change in all endpoints where we are using usernames.**
- For [user/details](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerUsers.js#L299), we might have to get the user record from ES and check if the token is valid before returning the credentials.
- [This query](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerStats.js#L78) could be a get or something with a filter on the start/end date.

**Some questions:**

- Can `admin`s only edit their own tasks? 
- How is `_key` [created](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerTasks.js#L301)? What ids are required in a geojson?
- We are [checking for `request.yar` along with `request.session`](https://github.com/osmlab/to-fix-backend/blob/f0209dbb2cc897e604897c89f375b770ba344fca/src/controllers/ControllerUsers.js#L55). How does this work? Do we need it?

cc: @rodowi @aaronlidman

Commit:
add process.env in config.js file

Commit:
machine update a task

Commit:
merge functions

Commit:
error in case the user is empty

Commit:
fixe response message

Commit:
change handle osmuser by user id

Commit:
update stats

Rub21 commented:
@ajithranka @geohacker @kepta  :
> Is it ok to have these hard-coded passwords here and here.

That should be a strong password setting by Environment variable!!

> Creating new tasks, updating or deleting tasks requires superadmin or admin privileges. Do we need the machine role to also be able to handle tasks?

- only `superadmin` and  `admin` can create a task https://github.com/osmlab/to-fix-backend/blob/2eb8c713ad290787d09d7c5c0097f521f4a9fc18/src/routes/routesTasks.js#L31

- only `superadmin`, `admin`, and `machine` can update a task  https://github.com/osmlab/to-fix-backend/blob/2a7a91e3a9c59c322e0db42616135cd341649c8b/src/routes/routesTasks.js#L56

> Should we have a strict check on uploaded files being .geojson? The file could be a .json as well

the file should be  only Geojson

> Do we need to catch errors here?
yes, this is ready!

> When updating a task, isCompleted should not be mandatory. Also, it shouldn't be hard-coded in the front-end.

I Don't understand here?

> The block of code starting from here which reads the uploaded file, validates it and updates the database could be broken into a function. It's repeated in both createTask and updateTask.
Same thing with this block.

you are right, I I've fixed it in just one function.

> Some checks might be useful when parsing the OSM user XML. This might break, for instance, if result.osm is empty.

The error is catch if the result is empty.,  However, I've fixed it. 

We probably need to remove this. Sets the user role to superadmin if the response is empty.

I've changed the name of the variable, it should be `exist` , it never will be empty.  only `true` or `false`.

> Do we need this userExists global? Can we limit it to the callback?

I need it  for the next `defer`

> Grammatical change in this error message: Bad authentication -- > Authentication failure.

fixed!

> Get user by id instead of username. Usernames can change and using string match on elastic search can be avoided (it will return many results). Same with changeRole. We should avoid relying on usernames everywhere. Let's use user ids as far as possible. This will require change in all endpoints where we are using usernames.

done!


> For user/details, we might have to get the user record from ES and check if the token is valid before returning the credentials.

This already check the authentication, we don't need to do it again

> This query could be a get or something with a filter on the start/end date.

done!


Still there is issues to fix, this coming week i will do it.

Commit:
user admin can update their own task

Commit:
rename funcion and variable

Rub21 commented:
>Can admins only edit their own tasks?

Yes, this is necessary, for that I've implemented it. done!! 

>How is _key created? What ids are required in a geojson?

_key is the hash of a Geojson feature,  and that is also the id for each item.

Commit:
remove request.session

Rub21 commented:
> We are checking for request.yar along with request.session. How does this work? Do we need it?

Since we are using token that is not necessary.  we just need the `yar`  to get the token from OSM response.

Commit:
clean up

Rub21 commented:
After talk with @ajithranka  we are ready to merge this to Master,
